ts/WoltLabSuite/Core/Ui/ScrollSpy.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * Executes provided callback if scroll threshold is reached. Usable to determine
+ * if user reached the bottom of an element to load new elements on the fly.
+ *
+ * If you do not provide a value for `reference` and `target` it will assume you're
+ * monitoring page scrolls, otherwise a valid HTMLElement must be provided.
+ *
+ * If reference is `null` the entire window will be the reference.
+ *
+ * @author  Joshua Ruesweg
+ * @copyright  2001-2021 WoltLab GmbH
+ * @license  GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @module  WoltLabSuite/Core/Ui/ScrollSpy
+ * @since 5.5
+ */
+
+type Callback = () => void;
+
+export class ScrollSpy {
+  private threshold: number;
+  private callback: Callback;
+  private reference?: HTMLElement;
+  private target?: HTMLElement;
+
+  public constructor(threshold: number, callback: Callback, target?: HTMLElement, reference?: HTMLElement) {
```

Comment:
Does it make sense to only provide one of `target` and `reference`?

If not:
```suggestion
  public constructor(threshold: number, callback: Callback);
  public constructor(threshold: number, callback: Callback, target: HTMLElement, reference: HTMLElement);
  public constructor(threshold: number, callback: Callback, target?: HTMLElement, reference?: HTMLElement) {
```
(untested)

Reply:
Yes. The common usage is to apply a `target` but not a `reference`.